### DISCUSSION TITLE
Issues with RunPCA on Seurat object with BPCells matrix directory

### DISCUSSION THREAD

**Discussion by fingeram** [satijalab/seurat#9327] [General]
Hi everyone, 

I am having problems with running dimensionality reduction on my seurat object created from a .h5ad file using BPCells. I checked the count matrix and everything looks completely fine (no NAs, values >=0). However, once I reach the step of RunPCA in my pipeline and then check the result using an Elbow plot, something looks very wrong and I simply cannot figure out what is the cause of it. Like why is the standard deviation so large and why does it drop after the 1st PC? Also, when trying to run clustering after RunPCA, it returns an error (see below).

I have the same problem if I subset my seurat object to reduce its size and transform the BPCells count matrix into a dgCMatrix...PCA embeddings and ElbowPlot still make no sense. 

**#import raw data from h5ad file**
raw <- open_matrix_anndata_hdf5(path="/novo/projects/departments/compbio/sysbio/Projects/mouse_liver_models/single_cell_and_nuclei/concatenated.dir/concatenated.h5ad") 

**#write matrix directory for on-disk storage** 
write_matrix_dir(mat = raw, dir = "/novo/projects/shared_projects/liver_biology_colab/people/aqnf/mouse_sc_sn_AQNF_June24/BPcells/mouse_counts")

**#load data from matrix directory to generate seurat object**
raw.mat <- open_matrix_dir(dir = "/novo/projects/shared_projects/liver_biology_colab/people/aqnf/mouse_sc_sn_AQNF_June24/BPcells/mouse_counts")

**#merge meta data** 
meta <- merge(x= metadata_BSCK, y= metadata_CPDM, by.x = "LibraryID", by.y = "library_id", all.y=T)
rownames(meta) <- meta$barcode

**# generate Seurat object from on-disk matrix**
sobj <- CreateSeuratObject(counts = raw.mat, meta.data = meta)

![image](https://github.com/user-attachments/assets/1b53c06e-a2f9-499b-b111-86720a65083a)

vals <- as(sobj[["RNA"]]$counts[1:500,1:500], "dgCMatrix")
anyNA(vals) 
all(vals >= 0) 
any(vals == 0)

![image](https://github.com/user-attachments/assets/267c17f1-8fab-4f63-9a67-f4188388d947)

**# process seurat object**
sobj <- NormalizeData(sobj)
sobj <- FindVariableFeatures(sobj)
sobj <- ScaleData(sobj)
sobj <- RunPCA(sobj)
ElbowPlot(sobj.sketch, ndims = 50, reduction = 'pca')

![cfbfcf90-e68a-4417-b47f-a56927de04d0](https://github.com/user-attachments/assets/8e3f7b2c-3af0-4a90-9c76-5f68752e8e0e)

> sobj <- FindNeighbors(sobj, dims = 1:50)
Computing nearest neighbor graph
Computing SNN
Error: std::bad_alloc

**Comment by zoezhao1030** [comment]
Same issue!

**Comment by longmanz** [comment]
Hi, thank you for bringing up this issue. Have you tried to remove cells with low UMI counts (e.g., nCount_RNA <=200)? It feels like some kinds of QC problems with the data.

**Reply by zoezhao1030** [reply]
My data is actually Xenium data and I only removed cells with UMI=0. I used SCT then ran RunPCA, I used the same input and tried these two steps several times independently -sometimes the output of the elbowplot is weird like the ones mentioned above, sometimes the output is normal with expected “elbow” shape.  In other words, the results are inconsistent despite the fact I used the exactly same input for SCT then RunPCA. (Ps I’m using Seurat V5)

**Reply by longmanz** [reply]
Hi,  thank you for the information. 
First, we no longer recommend running SCT on spatial datasets (see debate brought up here: https://genomebiology.biomedcentral.com/articles/10.1186/s13059-024-03241-7). You may consider using standard log-normalization or even not doing library-size normalization (only do log-transform). 
Second, it is probably not a good idea to keep cells with very low UMI. A minimal level of QC is required to avoid unstable decomposition in PCA.

**Reply by zoezhao1030** [reply]
Thank you for your suggestion! May I ask another question related to spatial analysis? What would be a recommended workflow for subset analysis? If I use LogNorm and regular preprocessing then use CCA to integrate the data and annotate my cell types. Then I want to subset specific cell type for subtype analysis, should I redo preprocessing+runPCA, then no reintegration and directly rerun FindNeighbors and FindClusters on already existed CCA result?  Thanks.

**Reply by longmanz** [reply]
It really depends on your purpose. In general you should not need to redo preprocessing or re-integrate your subset. You should simply re-run FindNeighbors and FindClusters for your subset. Unless you believe there are still uncorrected batch effect in your existing CCA results, then you may consider re-doing everything for your subset.

**Reply by zoezhao1030** [reply]
Got it. That makes sense. Thank you very much for your support and time!
